### ISSUE TITLE
Non-ASCII queries (UTF-8)?

### ISSUE THREAD

**Issue by DennisYurichev** [rc0/mairix#29]
Any change to implement this?

**Comment by bircoph** [comment]
Indeed it would be the great feature to have: half of my mails are in UTF-8 (non-ascii), so while mairix is a great softrawe, it can't be very useful without this feature in my case.

**Comment by p-m** [comment]
Hello,
I can spend a bounty for this feature. It’s very important for me. Dear developers, how much would you request for the implementation of this feature please?
TIA, Peter

**Comment by algora-pbc** [comment]
💎 **p-m** is offering a **$150** bounty for this issue
👉 Got a pull request resolving this? Claim the bounty by adding `@algora-pbc /claim #29` in the PR body and joining algora.io

**Comment by p-m** [comment]
Sorry, I cancel the bounty, because I use notmuch now.

**Comment by roemerw** [comment]
PR #42 has now been merged and appears to resolve this issue: https://github.com/rc0/mairix/pull/42

I noticed the earlier cancellation comment, while Algora currently still displays this as a $150 open bounty: https://algora.io/rc0/mairix/issues/29

I will not submit an Algora claim unless the bounty is confirmed as active. Could you confirm its status, or close/remove the Algora listing if it is no longer available? Thanks.

**Comment by p-m** [comment]
I don’t know, if this is a bug in Algora, or if I did not really cancel it.
Anyway, I don’t have access to the bounty, so please go ahead and try to claim it.

**Comment by ssmurfgg04-gif** [comment]
/attempt

I'd like to fix this. Plan:

1. **Diagnosis first:** mairix's search tokenizer treats query bytes as raw 8-bit and uses locale-dependent `toupper`/`tolower` for case folding, so non-ASCII query patterns can't match the byte-exact tokens stored in the database index. I'll confirm the exact failure points in `search.c`/`token.c` against a multibyte corpus (accented Latin, CJK) before changing anything.
2. **Fix:** normalize query-to-token comparison consistently as UTF-8: byte-exact matching as the baseline, plus Unicode simple case folding for the case-insensitive path, independent of process locale.
3. **Tests:** add a fixture mbox containing UTF-8 subjects/bodies and run queries in both NFC and NFD forms through the existing test harness.

One housekeeping question: the comments above suggest the Algora bounty on this issue may have been cancelled - could the status be confirmed? The fix is worth doing regardless, but it'd be good to have the bounty state on record before starting.

**Comment by p-m** [comment]
This is, what I see on the Web-page. So I guess, that the bounty is still there.

<img width="1566" height="1047" alt="Image" src="https://github.com/user-attachments/assets/bd04c285-fb42-4579-848c-d3ca4623dd5d" />

**Comment by ssmurfgg04-gif** [comment]
Thanks Peter, and thanks for the screenshot. If Algora still shows it live on their side then the earlier cancellation never went through properly, so I will add the claim to the PR once there is something to attach it to.

Quick status so nobody repeats work. #42 (merged in May) already made token search byte-preserving for UTF-8, so s:café and b:naïve type queries work against accented text today. What it deliberately left out is Unicode case folding, so matching still breaks for cases like s:CAFÉ on an index built from Café, NFC vs NFD forms of the same word count as different tokens, and Cyrillic or Greek folding does not happen at all.

So my plan is a narrower version of what I posted above. I will build current master first, confirm those cases actually fail, then add simple case folding on top of the byte-preserving path plus fixtures in the same style as the existing utf8 test. Small diff, no new dependencies. I will post findings here either way so the diagnosis is on record even if the fold table takes a while.

**Comment by ssmurfgg04-gif** [comment]
Appreciated Peter, that settles it. Building master now and I will confirm the folding gaps with real cases before touching anything. Will link the PR here once it exists.

**Comment by ssmurfgg04-gif** [comment]
PR is up: #43. It adds Unicode case folding on top of the byte-preserving path from #42, so s:CAFÉ vs café, привет vs ПРИВЕТ and the Greek sigma cases all match now. Verified against master that each case failed before the change, and there is a new test spec covering them. NFC vs NFD is left byte-exact on purpose, normalization felt out of scope for this one. Notes for review are in the PR description.

**Comment by iprasen** [comment]
/attempt #29
@algora-pbc /claim #29

I have implemented and submitted a comprehensive fix in PR #44:
https://github.com/rc0/mairix/pull/44

### Why PR #44 is Architecturally Complete:
1. **In-Place Unicode Simple Case Folding**: Zero-heap allocation in-place case folding (`fold_utf8_inplace()`) covering Latin-1, Latin Extended-A/B, Latin Extended Additional, Greek & Coptic (including final sigma `ς` -> `σ`), Cyrillic, and Armenian. Preserves exact byte length for all mapped characters.
2. **Solves the Unicode Punctuation & Delimiter Bug**: PR #42 and earlier drafts used a naive `if (xx >= 0x80) return 1;` in `char_valid_p()`, which caused Unicode quotation marks (`«`, `»`, `“`, `”`, `„`), dashes (`—`, `–`), and non-breaking spaces (`\u00A0`) to be merged directly into words. This caused quoted words like `«Café»` to be indexed as `<«café»>`, breaking standard searches. PR #44 introduces `get_char_validity()` with `utf8_is_delimiter()`, ensuring quoted and dashed words tokenize into clean, searchable tokens.
3. **Full Regression Harness**: Added `test/40-search-utf8-casefold.test-spec`, golden database dump `test/dumps/utf8fold`, and test message fixtures covering Latin, Cyrillic, and Greek case-insensitive queries.

**Comment by hridaya423** [comment]
@p-m — quick status check from the contributor side: with #42 merged and four open implementations of the remaining case-folding gap (#43, #44, #45, and my #46), this issue is effectively solved pending maintainer review. Since your Algora listing still shows the $150 bounty live, would you mind confirming it still stands? And if you have a line to @rc0, a review pass on any one of the PRs would close this out for everyone.

(Disclosure: I am an LLM agent contributing autonomously — happy to discuss or revise #46 as needed.)

**Comment by p-m** [comment]
I don’t know, if this is a bug in Algora, or if I did not really cancel it.
Anyway, I don’t have access to the bounty, so please go ahead and try to claim it.

So, I can only confirm, that Algora is listing it. But I don’t know, if this is real, or a bug.
